src/data/nat/basic.lean

```diff
@@ -1373,6 +1386,25 @@ by rw [nat.mul_add_mod, nat.mod_eq_of_lt h]
 lemma pred_eq_self_iff {n : ℕ} : n.pred = n ↔ n = 0 :=
 by { cases n; simp [(nat.succ_ne_self _).symm] }
 
+/-- A recursor that splits `n` into a multiple of `b` and a remainder `r` -/
+def div_mod_cases {C : ℕ → Sort*} (n b : ℕ) (hb : b ≠ 0)
+  (hm : ∀ (k : ℕ) (r < b), C (b * k + r)) : C n :=
+cast (congr_arg _ (div_add_mod n b)) (hm (n / b) (n % b) (mod_lt _ $ nat.pos_of_ne_zero hb))
+
+/-- Recursion principle based on the div and modulus decomposition against some base `b : ℕ`:
+given a map `C k → C (k + b)` for each `k`, and the construction of `C r` for all `r < b`,
+one can construct `C n` for any `n : ℕ`.  -/
+@[elab_as_eliminator]
+def div_mod_rec {C : ℕ → Sort*} (n b : ℕ) (hb : b ≠ 0) (hr : ∀ r < b, C r)
+  (hm : ∀ k, C k → C (k + b)) : C n :=
```

Comment:
The naming feels a bit weird here, given neither declaration has any mention of `div` or `mod` in its type; and it seems odd to name things after how they are proved.